File: src/main/java/link/infra/packwiz/installer/InstallWindow.java
package link.infra.packwiz.installer;

import java.awt.BorderLayout;
import java.awt.Component;
import java.awt.EventQueue;
import java.awt.GridBagConstraints;
import java.awt.GridBagLayout;

import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JOptionPane;
import javax.swing.JPanel;
import javax.swing.JProgressBar;
import javax.swing.UIManager;
import javax.swing.border.EmptyBorder;
import java.awt.event.ActionListener;
import java.awt.event.ActionEvent;

public class InstallWindow implements IUserInterface {

	private JFrame frmPackwizlauncher;
	private String title = "Updating modpack...";

	@Override
	public void show() {
		EventQueue.invokeLater(new Runnable() {
			public void run() {
				try {
					UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
					InstallWindow.this.initialize();
					InstallWindow.this.frmPackwizlauncher.setVisible(true);
				} catch (Exception e) {
					e.printStackTrace();
				}
			}
		});
	}

	/**
	 * Initialize the contents of the frame.
	 */
	private void initialize() {
		frmPackwizlauncher = new JFrame();
		frmPackwizlauncher.setTitle(title);
		frmPackwizlauncher.setBounds(100, 100, 493, 95);
		frmPackwizlauncher.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		frmPackwizlauncher.setLocationRelativeTo(null);
		
		JPanel panel = new JPanel();
		panel.setBorder(new EmptyBorder(10, 10, 10, 10));
		frmPackwizlauncher.getContentPane().add(panel, BorderLayout.CENTER);
		panel.setLayout(new BorderLayout(0, 0));
		
		JProgressBar progressBar = new JProgressBar();
		progressBar.setValue(50);
		panel.add(progressBar, BorderLayout.CENTER);
		
		JLabel lblProgresslabel = new JLabel("Loading...");
		panel.add(lblProgresslabel, BorderLayout.SOUTH);
		
		JPanel panel_1 = new JPanel();
		panel_1.setBorder(new EmptyBorder(0, 5, 0, 5));
		frmPackwizlauncher.getContentPane().add(panel_1, BorderLayout.EAST);
		GridBagLayout gbl_panel_1 = new GridBagLayout();
		panel_1.setLayout(gbl_panel_1);
		
		JButton btnOptions = new JButton("Options...");
		btnOptions.setAlignmentX(Component.CENTER_ALIGNMENT);
		GridBagConstraints gbc_btnOptions = new GridBagConstraints();
		gbc_btnOptions.gridx = 0;
		gbc_btnOptions.gridy = 0;
		panel_1.add(btnOptions, gbc_btnOptions);
		
		JButton btnCancel = new JButton("Cancel");
		btnCancel.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent arg0) {
				//updateManager.cleanup();
				frmPackwizlauncher.dispose();
			}
		});
		btnCancel.setAlignmentX(Component.CENTER_ALIGNMENT);
		GridBagConstraints gbc_btnCancel = new GridBagConstraints();
		gbc_btnCancel.gridx = 0;
		gbc_btnCancel.gridy = 1;
		panel_1.add(btnCancel, gbc_btnCancel);
	}

	@Override
	public void handleException(Exception e) {
		JOptionPane.showMessageDialog(null, e.getMessage(), title, JOptionPane.ERROR_MESSAGE);
	}
	
	@Override
	public void setTitle(String title) {
		this.title = title;
		if (frmPackwizlauncher != null) {
			EventQueue.invokeLater(new Runnable() {
				public void run() {
					InstallWindow.this.frmPackwizlauncher.setTitle(title);
				}
			});
		}
	}

}

File: src/main/kotlin/link/infra/packwiz/installer/ui/gui/InstallWindow.kt
package link.infra.packwiz.installer.ui.gui

import link.infra.packwiz.installer.Msgs
import link.infra.packwiz.installer.ui.data.InstallProgress
import java.awt.BorderLayout
import java.awt.Component
import java.awt.GridBagConstraints
import java.awt.GridBagLayout
import javax.swing.*
import javax.swing.border.EmptyBorder

class InstallWindow(private val handler: GUIHandler) : JFrame() {
	private var lblProgresslabel: JLabel
	private var progressBar: JProgressBar
	private var btnOptions: JButton
	private val btnCancel: JButton
	private val btnOk: JButton
	private val buttonsPanel: JPanel

	init {
		setBounds(100, 100, 493, 95)
		// Works better with tiling window managers - there isn't any reason to change window size currently anyway
		isResizable = false
		defaultCloseOperation = EXIT_ON_CLOSE
		setLocationRelativeTo(null)

		// Progress bar and loading text
		add(JPanel().apply {
			border = EmptyBorder(10, 10, 10, 10)
			layout = BorderLayout(0, 0)

			progressBar = JProgressBar().apply {
				isIndeterminate = true
			}
			add(progressBar, BorderLayout.CENTER)

			lblProgresslabel = JLabel(Msgs.hintMore(Msgs.loading()))
			add(lblProgresslabel, BorderLayout.SOUTH)
		}, BorderLayout.CENTER)

		// Buttons
		buttonsPanel = JPanel().apply {
			border = EmptyBorder(0, 5, 0, 5)
			layout = GridBagLayout()

			btnOptions = JButton(Msgs.hintMore(Msgs.optionalMods())).apply {
				alignmentX = Component.CENTER_ALIGNMENT

				addActionListener {
					text = Msgs.hintMore(Msgs.loading())
					isEnabled = false
					handler.optionsButtonPressed = true
				}
			}
			add(btnOptions, GridBagConstraints().apply {
				gridx = 1
				gridy = 0
			})

			btnCancel = JButton(Msgs.cancel()).apply {
				addActionListener {
					isEnabled = false
					handler.cancelButtonPressed = true
				}
			}
			add(btnCancel, GridBagConstraints().apply {
				gridx = 1
				gridy = 1
			})
		}

		btnOk = JButton(Msgs.continueText()).apply {
			addActionListener {
				handler.okButtonPressed = true
			}
		}
		add(buttonsPanel, BorderLayout.EAST)
	}

	fun displayProgress(progress: InstallProgress) {
		if (progress.hasProgress) {
			progressBar.isIndeterminate = false
			progressBar.value = progress.progress
			progressBar.maximum = progress.progressTotal
		} else {
			progressBar.isIndeterminate = true
			progressBar.value = 0
		}
		lblProgresslabel.text = progress.message
	}

	fun disableOptionsButton(hasOptions: Boolean) {
		btnOptions.apply {
			text = if (hasOptions) { Msgs.hintMore(Msgs.optionalMods()) } else { Msgs.noOptionalMods() }
 			isEnabled = false
		}
	}

	fun showOk(hideCancel: Boolean) {
		if (hideCancel) {
			buttonsPanel.add(btnOk, GridBagConstraints().apply {
				gridx = 1
				gridy = 1
			})
			buttonsPanel.remove(btnCancel)
		} else {
			buttonsPanel.add(btnOk, GridBagConstraints().apply {
				gridx = 0
				gridy = 1
			})
		}
		buttonsPanel.revalidate()
	}

	fun hideOk() {
		buttonsPanel.remove(btnOk)
		if (!buttonsPanel.components.contains(btnCancel)) {
			buttonsPanel.add(btnCancel, GridBagConstraints().apply {
				gridx = 1
				gridy = 1
			})
		}
		buttonsPanel.revalidate()
	}

	fun timeoutOk(remaining: Long) {
		btnOk.text = Msgs.countdown(Msgs.continueText(), remaining)
	}
}